### ISSUE TITLE
Support GeoPoint in buckets_path references

### ISSUE THREAD

**Issue by talevy** [elastic/elasticsearch#39957]
**Feature Request**

Support geo_point references in `buckets_path` for pipeline aggregations

**Example**

It would be nice to enrich a bucket response with more information. In this case, 
it would be nice if there were a bucket_script aggregation alongside a geotile_grid
that returned a geo_shape representation of the geo bucket keys. Currently, geo-grid
aggregation's key values are represented as geo-points.

This will allow users to leverage scripts as part of the request so that dedicated
flags for custom/extra representations are not necessary as is requested in #39810.

```
POST /museums/_search?size=0
{
  "aggregations": {
    "geogrid": {
      "geotile_grid": {
        "field": "location",
        "precision": 8
      },
      "aggs": {
        "boundingBoxOfGridBucket": {
          "bucket_script": {
            "buckets_path": {
              "key": "_key"
            },
            "script": "..."
          }
        }
      }
    }
  }
}
```

**Exception/Issue**

```
{
  "error": {
    "root_cause": [],
    "type": "search_phase_execution_exception",
    "reason": "",
    "phase": "fetch",
    "grouped": true,
    "failed_shards": [],
    "caused_by": {
      "type": "aggregation_execution_exception",
      "reason": "buckets_path must reference either a number value or a single value numeric metric aggregation, got: org.elasticsearch.common.geo.GeoPoint"
    }
  },
  "status": 503
}
```

**Comment by nyurik** [comment]
@talevy thanks, i really like this solution, as it makes the result far more flexible than just getting a predetermined geo-bucket shape.  When you mentioned `GeoPoint` above, will that also cover the usecase of the entire bucket shape (rectangular/hex/...), or will it only support the centerpoint of the bucket?

**Comment by talevy** [comment]
> When you mentioned GeoPoint above, will that also cover the usecase of the entire bucket shape (rectangular/hex/...), or will it only support the centerpoint of the bucket?

yeah... to be determined. I remember we chose to stick with returning GeoPoint for the key value... slightly regret it. Will know more about how to circumvent this for sanity in the scripts when an implementation is investigated.

**Comment by nyurik** [comment]
Well, 7.x hasn't shipped yet, maybe we can still make the simple change. The bigger question is what should be returned, or if there should be another method `getShape()` on the bucket...

**Comment by talevy** [comment]
@nyurik, I do not feel comfortable rushing in a change like this before we understand the problem a bit more.

For example, it looks like each MultiBucketAggregation has control over what is returned for these special values. I did some digging and will include a diff to InternalGeoHashGridBucket/InternalGeoTileGridBucket that can be done to bypass the existing `getKey()`

```diff
--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/InternalGeoHashGridBucket.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/InternalGeoHashGridBucket.java
 public class InternalGeoHashGridBucket extends InternalGeoGridBucket<InternalGeoHashGridBucket> {
     InternalGeoHashGridBucket(long hashAsLong, long docCount, InternalAggregations aggregations) {
@@ -52,4 +56,30 @@ public class InternalGeoHashGridBucket extends InternalGeoGridBucket<InternalGeo
     public GeoPoint getKey() {
         return GeoPoint.fromGeohash(hashAsLong);
     }
+
+    @Override
+    public Object getProperty(String containingAggName, List<String> path) {
+        if (path.isEmpty()) {
+            return this;
+        }
+        Aggregations aggregations = getAggregations();
+        String aggName = path.get(0);
+        if (aggName.equals("_count")) {
+            if (path.size() > 1) {
+                throw new InvalidAggregationPathException("_count must be the last element in the path");
+            }
+            return getDocCount();
+        } else if (aggName.equals("_key")) {
+            if (path.size() > 1) {
+                throw new InvalidAggregationPathException("_key must be the last element in the path");
+            }
+            return getKey(); // can be modified!!!!!!!!!
+        }
+        InternalAggregation aggregation = aggregations.get(aggName);
+        if (aggregation == null) {
+            throw new InvalidAggregationPathException("Cannot find an aggregation named [" + aggName + "] in [" + containingAggName
+                + "]");
+        }
+        return aggregation.getProperty(path.subList(1, path.size()));
+    }
 }
```

we can modify where I commented "can be modified!!!" so that getKey() is not used, but instead `hashAsLong` is returned (assuming the requirement for numeric values is still enforced)

then the script can encode `hashAsLong` to a geojson geometry.

does that sound reasonable @polyfractal? My naive views from skimming make it look like this would just work. I will investigate later

**Comment by talevy** [comment]
I've updated the title and description of this issue to reflect the proposed implementation

**Comment by nyurik** [comment]
@talevy agree, makes sense to think it through. My main concern is to be able to get the whole `LineString` out, as that seems to be more valuable for some visualizations.  Also, note that LineString is not closed by default, and to close it, the first and last point must be identical (e.g. a rectangle is encoded with 5 points). Lastly, we `may` want to allow users to do more complex computations, e.g. ask ES to project geopoints using some specific projection, extract the zoom level, calculate bucket surface area, etc, but this is clearly something for later.

**Comment by talevy** [comment]
makes sense to want to do those things, that is even more reason why I think it is good to start with scripts. If we see usage in these scripts that will be provided in the docs, then we can start thinking about how to introduce these features internally.

**Comment by polyfractal** [comment]
> does that sound reasonable @polyfractal? My naive views from skimming make it look like this would just work. I will investigate later. I think the alternative is to natively support GeoPoint, which isn't actually what we want to do in this use-case. So maybe this lighter solution is actually what we want

I think that would technically work, although I don't think any of the other aggs override `getProperty()` to re-define how the special paths work.  E.g. it's odd that GeoHash returns `x` for `getKey()` method, but `y` for `_key` (I understand there's here history here, just from a theoretical point of view).

As you said, the other option is to support GeoPoint directly.  I think that's perhaps not so outlandish.  Right now `bucket_script` treats all return values as `Double`, but there are other requests to [handle strings, etc](https://github.com/elastic/elasticsearch/issues/36642).  BucketScript mainly [takes the resolved values](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketScriptPipelineAggregator.java#L101), does some `gap_policy` checking and then sticks them in a map for the script to consume.

If `bucket_script` were tweaked to accept Objects (and figure out something to do with the gap_policy situation), and [BucketHelpers was tweaked](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java#L158-L186) to allow GeoPoints/Strings/whatever the `bucket_script` would be allowed to consume whatever type it wanted.

The sticky bits are all the interactions with other pipelines: what happens when someone tries to use a geopoint in a `derivative` agg?  We'd want something to catch those instances so we could use a nice error, instead of a bad-cast exception.

**Comment by talevy** [comment]
more digging to be done for sure! thanks for the links. one other thing to point out is that the geo-point itself is not enough, right? one would have to also add the `precision` along the geo-point to be able to decode it into a shape. Taking a look at bucket_script... it does not allow for additional static script parameters. This parameter would have to be duplicated and in-sync with the geo-grid aggregation value.

**Comment by polyfractal** [comment]
Oops, you're right. ++ 

I was thinking we'd return the whole geo bucket object and allow the user to get the key or hash from that.  But I confused myself when writing it up, and I'm not entirely sure that's a viable option either :)

It's a similar for other complex objects too.  A user wanting to use a stats or percentiles agg has to request all the relevant stats/percentiles as different paths, instead of just getting the raw agg itself to request from.  It makes sense for other pipelines... but for a script, maybe we could just give them the object itself? 

All that with a grain of salt, just a half-baked vague idea in my head... it might be a terrible idea :)

>  Taking a look at bucket_script... it does not allow for additional static script parameters.

The `script` parameter uses the global script parser, so it can either accept a string, or the longform `script: {}` object which can include static params.  _Should_ work, although I'm not sure I've personally tried it :)

But yeah, making the user pass in the precision is non-ideal regardless.  I'd prefer if we didnt need to rely on the user for that.

#### Third option
To throw out another option, we could also add a new special path (`_hash_key` or something) which is only relevant to the geo hash aggs.  Not super ideal, but not all the paths work universally anyway (`_key` doesn't work on metric aggs, etc)

**Comment by talevy** [comment]
Zach and I spoke off-github and came to the conclusion that we think a good starting place is to explore adding support for String-valued parameters for bucket-scripts and expose a new special accessor called `_key_as_string`.

Then the accessor scope to buckets would be one-to-one with the Bucket interface's (getKey, getKeyAsString, getDocCount)

**Comment by wchaparro** [comment]
Stale issue, not going to address. Closing as won't do.

Reopen if prioritization is incorrect.
